# 1. Resumen de los datos: dimensiones y estructuras

In [98]:
import pandas as pd

In [123]:
mainpath = "/Users/irene/Documents/GitHub/python-ml-course/datasets"  #Ruta ficheros
filename1 = "titanic/titanic3.csv" #Fichero a abrir
fullpath1 = mainpath + "/" + filename1 #Ruta completa

df1 = pd.read_csv(fullpath1)

### Cabeceras

Para verificar si los datos se han cargado bien.

Puedes especificar en el paréntesis cuántos elementos quieres que lea, por def: 5.

In [100]:
df1.head(10)

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5       1         1                              Anderson, Mr. Harry    male   
6       1         1                Andrews, Miss. Kornelia Theodosia  female   
7       1         0                           Andrews, Mr. Thomas Jr    male   
8       1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9       1         0                          Artagaveytia, Mr. Ramon    male   

       age  sibsp  parch    ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0     24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2    113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2    113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2    113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2    113781  151.5500  C22 C26        S  NaN    NaN   
5  48.0000      0      0     19952   26.5500      E12        S    3    NaN   
6  63.0000      1      0     13502   77.9583       D7        S   10    NaN   
7  39.0000      0      0    112050    0.0000      A36        S  NaN    NaN   
8  53.0000      2      0     11769   51.4792     C101        S    D    NaN   
9  71.0000      0      0  PC 17609   49.5042      NaN        C  NaN   22.0   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON  
5                     New York, NY  
6                       Hudson, NY  
7                      Belfast, NI  
8              Bayside, Queens, NY  
9              Montevideo, Uruguay

### Colas

Para leer los últimos (5) elementos

In [101]:
df1.tail(10)

pclass  survived                                     name     sex   age  \
1299       3         0                      Yasbeck, Mr. Antoni    male  27.0   
1300       3         1  Yasbeck, Mrs. Antoni (Selini Alexander)  female  15.0   
1301       3         0                     Youseff, Mr. Gerious    male  45.5   
1302       3         0                        Yousif, Mr. Wazli    male   NaN   
1303       3         0                    Yousseff, Mr. Gerious    male   NaN   
1304       3         0                     Zabour, Miss. Hileni  female  14.5   
1305       3         0                    Zabour, Miss. Thamine  female   NaN   
1306       3         0                Zakarian, Mr. Mapriededer    male  26.5   
1307       3         0                      Zakarian, Mr. Ortin    male  27.0   
1308       3         0                       Zimmerman, Mr. Leo    male  29.0   

      sibsp  parch  ticket     fare cabin embarked boat   body home.dest  
1299      1      0    2659  14.4542   NaN        C    C    NaN       NaN  
1300      1      0    2659  14.4542   NaN        C  NaN    NaN       NaN  
1301      0      0    2628   7.2250   NaN        C  NaN  312.0       NaN  
1302      0      0    2647   7.2250   NaN        C  NaN    NaN       NaN  
1303      0      0    2627  14.4583   NaN        C  NaN    NaN       NaN  
1304      1      0    2665  14.4542   NaN        C  NaN  328.0       NaN  
1305      1      0    2665  14.4542   NaN        C  NaN    NaN       NaN  
1306      0      0    2656   7.2250   NaN        C  NaN  304.0       NaN  
1307      0      0    2670   7.2250   NaN        C  NaN    NaN       NaN  
1308      0      0  315082   7.8750   NaN        S  NaN    NaN       NaN

### Dimensión

nº filas x nº columnas

In [102]:
df1.shape

(1309, 14)

### Nombre columnas

In [103]:
df1.columns.values 

array(['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch',
       'ticket', 'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest'],
      dtype=object)

### Resumen de los estadísticos básicos de las variables numéricas

- count: nº de objetos, no cuenta NaNs.
- mean: promedio (suma de los valores / nº de valores)
- std: desviación estándar, cómo se desvian los valores respecto al promedio
- min/max: valor mínimo/máximo.
- 25%,50%,75%: cuantiles, aquellos valores que, ordenados de min a max, caen justo en esa posición respecto a la ordenación total.

In [104]:
df1.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881135     0.498854     0.385027   
std       0.837836     0.486055    14.413500     1.041658     0.865560   
min       1.000000     0.000000     0.166700     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000

### Tipo de dato de cada columna 

In [105]:
df1.dtypes #object=string

pclass         int64
survived       int64
name          object
sex           object
age          float64
sibsp          int64
parch          int64
ticket        object
fare         float64
cabin         object
embarked      object
boat          object
body         float64
home.dest     object
dtype: object

# 2. Missing values

Sabemos que faltan valores porque tenemos 1310 filas, y de la columna "Age" sólo tenemos 1046 datos, hay 264 que no tienen edad. Aunque esa función es solo para variables numéricas.

In [106]:
pd.isnull(df1["body"]) #como argumento podemos pasarle el nombre de las columnas, si no, procesa todo

# True -> valor nulo, NaN o None

0        True
1        True
2        True
3       False
4        True
        ...  
1304    False
1305     True
1306    False
1307     True
1308     True
Name: body, Length: 1309, dtype: bool

In [107]:
pd.notnull(df1["body"]) #Lo contrario

0       False
1       False
2       False
3        True
4       False
        ...  
1304     True
1305    False
1306     True
1307    False
1308    False
Name: body, Length: 1309, dtype: bool

In [108]:
pd.isnull(df1["body"]).values #Para convertirlo en un array

array([ True,  True,  True, ..., False,  True,  True])

In [109]:
pd.isnull(df1["body"]).values.ravel() #Para construir un array de 1 única fila

array([ True,  True,  True, ..., False,  True,  True])

In [110]:
pd.isnull(df1["body"]).values.ravel().sum() #Para sumar los valores de ese Array, suma los Trues (1)

1188

### Borrado de valores que faltan (falta el 1-3%)

In [111]:
df2 = df1  #Copia para no borrar todo con el ejemplo

#Caso menos drástico (All): borramos sólo si todos los valores son NaNs
df2.dropna(axis=0,how="all") #Borramos filas (eje X)

#Caso más drástico (Any): borramos sólo si 1 valor tiene NaN
df2.dropna(axis=1,how="any") # Borramos columnas (eje Y)

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex  sibsp  parch  ticket  
0     female      0      0   24160  
1       male      1      2  113781  
2     female      1      2  113781  
3       male      1      2  113781  
4     female      1      2  113781  
...      ...    ...    ...     ...  
1304  female      1      0    2665  
1305  female      1      0    2665  
1306    male      0      0    2656  
1307    male      0      0    2670  
1308    male      0      0  315082  

[1309 rows x 7 columns]

### Inferencia o cómputo de valores que faltan

Consiste en reemplazar los valores que faltan por otros valores -> 0, un string, el promedio del resto de valores... depende del contexto de la información.

In [128]:
df3 = df1 #Vamos a sobreescribir en un df nuevo

df3 = df3.fillna(0) #Cambiar por 0
df3 = df3.fillna("Desconocido") #Cambiar por String

# Reemplazar dependiendo de la columna:

df3["body"] = df3["body"].fillna(0) #Porque el campo es numérico
df3["home.dest"] = df3["home.dest"].fillna("Desconocido") #Porque el campo es un string

df3["age"] = df3["age"].fillna(df3["age"].mean()) #Por el promedio
df3["age"] = df3["age"].fillna(method="ffill") #Con el valor inmediatamente posterior ffill  = forward
df3["age"] = df3["age"].fillna(method="backfill") #Con el valor inmediatamente anterior backfill = bakward 

### Variables dummy 

Variables artificiales, para las variables categóricas -> creamos dos variables dummy para la columna "sex" -> 
columna male y columna female. Crear tantas columnas como categorías tenga.

In [146]:
df4 = df1 #Vamos a sobreescribir en un df nuevo

dummy_sex = pd.get_dummies(df4["sex"], prefix="sex")  #De la columna Sex, crea columnas con prefijo sex.
dummy_sex.head()

sex_female  sex_male
0           1         0
1           0         1
2           1         0
3           0         1
4           1         0

Eliminar la variable original, categórica, y añadir las nuevas columnas a la tabla.

In [147]:
column_name = df4.columns.values.tolist()
column_name

['pclass',
 'survived',
 'name',
 'sex',
 'age',
 'sibsp',
 'parch',
 'ticket',
 'fare',
 'cabin',
 'embarked',
 'boat',
 'body',
 'home.dest']

In [148]:
df4 = df4.drop(["sex"], axis=1)
df4

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

          age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0     29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1      0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2      2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3     30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4     25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
...       ...    ...    ...     ...       ...      ...      ...  ...    ...   
1304  14.5000      1      0    2665   14.4542      NaN        C  NaN  328.0   
1305  14.5000      1      0    2665   14.4542      NaN        C  NaN    NaN   
1306  26.5000      0      0    2656    7.2250      NaN        C  NaN  304.0   
1307  27.0000      0      0    2670    7.2250      NaN        C  NaN    NaN   
1308  29.0000      0      0  315082    7.8750      NaN        S  NaN    NaN   

                            home.dest  
0                        St Louis, MO  
1     Montreal, PQ / Chesterville, ON  
2     Montreal, PQ / Chesterville, ON  
3     Montreal, PQ / Chesterville, ON  
4     Montreal, PQ / Chesterville, ON  
...                               ...  
1304                              NaN  
1305                              NaN  
1306                              NaN  
1307                              NaN  
1308                              NaN  

[1309 rows x 13 columns]

In [149]:
column_name = df4.columns.values.tolist()
column_name

['pclass',
 'survived',
 'name',
 'age',
 'sibsp',
 'parch',
 'ticket',
 'fare',
 'cabin',
 'embarked',
 'boat',
 'body',
 'home.dest']

Para añadir (concatenar) las columnas con las variables dummy

In [152]:
df5 = pd.concat([df4, dummy_sex], axis=1)

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

          age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0     29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1      0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2      2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3     30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4     25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
...       ...    ...    ...     ...       ...      ...      ...  ...    ...   
1304  14.5000      1      0    2665   14.4542      NaN        C  NaN  328.0   
1305  14.5000      1      0    2665   14.4542      NaN        C  NaN    NaN   
1306  26.5000      0      0    2656    7.2250      NaN        C  NaN  304.0   
1307  27.0000      0      0    2670    7.2250      NaN        C  NaN    NaN   
1308  29.0000      0      0  315082    7.8750      NaN        S  NaN    NaN   

                            home.dest  sex_female  sex_male  
0                        St Louis, MO           1         0  
1     Montreal, PQ / Chesterville, ON           0         1  
2     Montreal, PQ / Chesterville, ON           1         0  
3     Montreal, PQ / Chesterville, ON           0         1  
4     Montreal, PQ / Chesterville, ON           1         0  
...                               ...         ...       ...  
1304                              NaN           1         0  
1305                              NaN           1         0  
1306                              NaN           0         1  
1307                              NaN           0         1  
1308                              NaN           0         1  

[1309 rows x 15 columns]

In [153]:
# Función para crear dummies:

def createDummies(df, var_name):
    dummy = pd.get_dummies(df[var_name], prefix=var_name)
    df = df.drop(var_name, axis=1)
    df = pd.concat([df, dummy], axis=1)
    return df


In [154]:
df6 = df1 

createDummies(df6, "sex")

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

          age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0     29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1      0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2      2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3     30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4     25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
...       ...    ...    ...     ...       ...      ...      ...  ...    ...   
1304  14.5000      1      0    2665   14.4542      NaN        C  NaN  328.0   
1305  14.5000      1      0    2665   14.4542      NaN        C  NaN    NaN   
1306  26.5000      0      0    2656    7.2250      NaN        C  NaN  304.0   
1307  27.0000      0      0    2670    7.2250      NaN        C  NaN    NaN   
1308  29.0000      0      0  315082    7.8750      NaN        S  NaN    NaN   

                            home.dest  sex_female  sex_male  
0                        St Louis, MO           1         0  
1     Montreal, PQ / Chesterville, ON           0         1  
2     Montreal, PQ / Chesterville, ON           1         0  
3     Montreal, PQ / Chesterville, ON           0         1  
4     Montreal, PQ / Chesterville, ON           1         0  
...                               ...         ...       ...  
1304                              NaN           1         0  
1305                              NaN           1         0  
1306                              NaN           0         1  
1307                              NaN           0         1  
1308                              NaN           0         1  

[1309 rows x 15 columns]

### 2. Crear una columna nueva

In [10]:
import pandas as pd

In [11]:
mainpath = "/Users/irene/Documents/GitHub/python-ml-course/datasets"  #Ruta ficheros
filename1 = "customer-churn-model/Customer Churn Model.txt" #Fichero a abrir
fullpath1 = mainpath + "/" + filename1 #Ruta completa

df1 = pd.read_csv(fullpath1)
df1.head()

State  Account Length  Area Code     Phone Int'l Plan VMail Plan  \
0    KS             128        415  382-4657         no        yes   
1    OH             107        415  371-7191         no        yes   
2    NJ             137        415  358-1921         no         no   
3    OH              84        408  375-9999        yes         no   
4    OK              75        415  330-6626        yes         no   

   VMail Message  Day Mins  Day Calls  Day Charge  ...  Eve Calls  Eve Charge  \
0             25     265.1        110       45.07  ...         99       16.78   
1             26     161.6        123       27.47  ...        103       16.62   
2              0     243.4        114       41.38  ...        110       10.30   
3              0     299.4         71       50.90  ...         88        5.26   
4              0     166.7        113       28.34  ...        122       12.61   

   Night Mins  Night Calls  Night Charge  Intl Mins  Intl Calls  Intl Charge  \
0       244.7           91         11.01       10.0           3         2.70   
1       254.4          103         11.45       13.7           3         3.70   
2       162.6          104          7.32       12.2           5         3.29   
3       196.9           89          8.86        6.6           7         1.78   
4       186.9          121          8.41       10.1           3         2.73   

   CustServ Calls  Churn?  
0               1  False.  
1               1  False.  
2               0  False.  
3               2  False.  
4               3  False.  

[5 rows x 21 columns]

In [8]:
action_columns = ["Day Mins", "Night Mins"]
df1["Total Mins"] = df1["Day Mins"] + df1["Night Mins"] + df1["Eve Mins"]
df1["Total Mins"].head()

0    707.2
1    611.5
2    527.2
3    558.2
4    501.9
Name: Total Mins, dtype: float64

In [9]:
df1.head() #Se ha añadido una nueva columna

State  Account Length  Area Code     Phone Int'l Plan VMail Plan  \
0    KS             128        415  382-4657         no        yes   
1    OH             107        415  371-7191         no        yes   
2    NJ             137        415  358-1921         no         no   
3    OH              84        408  375-9999        yes         no   
4    OK              75        415  330-6626        yes         no   

   VMail Message  Day Mins  Day Calls  Day Charge  ...  Eve Charge  \
0             25     265.1        110       45.07  ...       16.78   
1             26     161.6        123       27.47  ...       16.62   
2              0     243.4        114       41.38  ...       10.30   
3              0     299.4         71       50.90  ...        5.26   
4              0     166.7        113       28.34  ...       12.61   

   Night Mins  Night Calls  Night Charge  Intl Mins  Intl Calls  Intl Charge  \
0       244.7           91         11.01       10.0           3         2.70   
1       254.4          103         11.45       13.7           3         3.70   
2       162.6          104          7.32       12.2           5         3.29   
3       196.9           89          8.86        6.6           7         1.78   
4       186.9          121          8.41       10.1           3         2.73   

   CustServ Calls  Churn?  Total Mins  
0               1  False.       707.2  
1               1  False.       611.5  
2               0  False.       527.2  
3               2  False.       558.2  
4               3  False.       501.9  

[5 rows x 22 columns]